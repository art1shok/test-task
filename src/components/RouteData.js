import React from 'react';
import {
  RouteContainer,
  RouteDataStyled,
  RouteDataContainer,
  RouteDescription
} from '../styled/RouteData.styled';

const handleLandingTime = (startHours, startMinutes, flightTimeHours, flightTimeMinutes) => {
  let Hours = startHours + flightTimeHours;
  let Minutes = startMinutes + flightTimeMinutes;

  while (Hours >= 24) {
    Hours -= 24;
  }

  if (Minutes > 60) {
    Minutes -= 60;
  }

  return {
    hours: Hours,
    minutes: Minutes
  };
};

export const RouteData = ({ way, date, duration, stops }) => {
  const hh = new Date(date).getHours();
  const mm = new Date(date).getMinutes();
  const flightHours = Math.floor(duration / 60);
  const flightMinutes = duration %= 60;

  const landingTime = handleLandingTime(hh, mm, flightHours, flightMinutes);

  return (
    <RouteContainer md={{ span: 4 }}>
      <RouteDataContainer>
        <RouteDescription>{way}</RouteDescription>
        <RouteDataStyled>
          {`${hh < 10 ? '0' + hh : hh}:${mm < 10 ? '0' + mm : mm} -
           ${landingTime.hours < 10 ? '0' + landingTime.hours : landingTime.hours}:${
            landingTime.minutes < 10 ? '0' + landingTime.minutes : landingTime.minutes}`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer md={{ span: 4 }}>
        <RouteDescription>В пути</RouteDescription>
        <RouteDataStyled>
          {`${flightHours}ч ${flightMinutes}м`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer md={{ span: 3 }}>
        <RouteDescription>
          {stops.length === 0 ? 'Без пересадок' : stops.length === 1 ? '1 пересадка' : `${stops.length} пересадки`}
        </RouteDescription>
        <RouteDataStyled>{stops.join(', ')}</RouteDataStyled>
      </RouteDataContainer>
    </RouteContainer>
  );
};
