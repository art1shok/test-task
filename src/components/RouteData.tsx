import React, { FC, useMemo } from 'react';
import {
  RouteContainer,
  RouteDataStyled,
  RouteDataContainer,
  RouteDescription,
} from '../styled/RouteData.styled';

interface Props {
  way: string;
  date: string;
  duration: number;
  stops: string[];
}

const handleLandingTime = (startHours: number, startMinutes: number,
  flightTimeHours: number, flightTimeMinutes: number) => {
  let Hours = startHours + flightTimeHours;
  let Minutes = startMinutes + flightTimeMinutes;

  while (Hours >= 24) {
    Hours -= 24;
  }

  if (Minutes > 60) {
    Minutes -= 60;
  }

  return {
    hours: Hours,
    minutes: Minutes,
  };
};

export const RouteData: FC<Props> = ({
  way, date, duration, stops,
}) => {
  const hh = new Date(date).getHours();
  const mm = new Date(date).getMinutes();
  const flightHours = Math.floor(duration / 60);
  const flightMinutes = duration % 60;

  const landingTime = handleLandingTime(hh, mm, flightHours, flightMinutes);

  const addZeroes = (time: number) => (time < 10 ? `0${time}` : time);

  const formattedTime = useMemo(() =>
    `${addZeroes(hh)}:${addZeroes(mm)} - ${addZeroes(landingTime.hours)}:${addZeroes(landingTime.minutes)}`,
  [hh, landingTime.hours, landingTime.minutes, mm]);

  const handleStops = (transfers: string[]) => {
    if (transfers.length === 0) {
      return 'Без пересадок';
    } if (transfers.length === 1) {
      return '1 пересадка';
    }
    return `${transfers.length} пересадки`;
  };

  const transfers = handleStops(stops);

  return (
    <RouteContainer>
      <RouteDataContainer>
        <RouteDescription className="way">{way}</RouteDescription>
        <RouteDataStyled className="time">
          {formattedTime}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer>
        <RouteDescription>В пути</RouteDescription>
        <RouteDataStyled>
          {`${flightHours}ч ${flightMinutes}м`}
        </RouteDataStyled>
      </RouteDataContainer>

      <RouteDataContainer>
        <RouteDescription className="stops">
          {transfers}
        </RouteDescription>
        <RouteDataStyled className="transfers">{stops.join(', ')}</RouteDataStyled>
      </RouteDataContainer>
    </RouteContainer>
  );
};
