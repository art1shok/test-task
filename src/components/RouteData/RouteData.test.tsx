import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import 'jest-styled-components';

import { RouteData } from './RouteData';
import {
  RouteContainer,
} from '../../styled/RouteData.styled';

configure({ adapter: new Adapter() });

const ticket = {
  origin: 'MAK',
  destination: 'MAV',
  date: '10/03/2017 07:29:46-0700',
  stops: [
    'JAK',
    'DAV',
  ],
  duration: 1200,
};

describe('should render RouteData component', () => {
  let component: any = null;
  beforeEach(() => {
    component = shallow(
      <RouteData
        way={`${ticket.origin} - ${ticket.destination}`}
        date={ticket.date}
        duration={ticket.duration}
        stops={ticket.stops}
      />,
    );
  });

  it('RouteData component renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should contain container', () => {
    const container = component.find(RouteContainer);
    expect(container.length).toBe(1);
  });

  it('should contain correct format way', () => {
    const route = 'MAK - MAV';
    const way = component.find('.way');
    expect(way.text()).toBe(route);
  });

  it('should contain correct format time', () => {
    const time = '17:29 - 13:29';
    const container = component.find('.time');
    expect(container.text()).toBe(time);
  });

  it('should contain transfers count', () => {
    const count = '2 пересадки';
    const container = component.find('.stops');
    expect(container.text()).toBe(count);
  });

  it('should contain transfers cities', () => {
    const transfers = 'JAK, DAV';
    const container = component.find('.transfers');
    expect(container.text()).toBe(transfers);
  });
});
