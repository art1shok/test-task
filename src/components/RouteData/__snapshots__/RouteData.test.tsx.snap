// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RouteData component RouteData component renders correctly 1`] = `ShallowWrapper {}`;
